text/0000-experimental-coroutines.md

```diff
@@ -0,0 +1,471 @@
+- Feature Name: `coroutines`
+- Start Date: 2017-06-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This is an **experimental RFC** for adding a new feature to the language,
+coroutines (also commonly referred to as generators). This RFC is intended to be
+relatively lightweight and bikeshed free as it will be followed by a separate
+RFC in the future for stabilization of this language feature. The intention here
+is to make sure everyone's on board with the *general idea* of
+coroutines/generators being added to the Rust compiler and available for use on
+the nightly channel.
+
+# Motivation
+[motivation]: #motivation
+
+One of Rust's 2017 roadmap goals is ["Rust should be well-equipped for writing
+robust, high-scale servers"][goal]. A [recent survey][survey] has shown that
+the biggest blocker to robust, high-scale servers is ergonomic usage of async
```

Comment:
My point about the cache is this: some coroutines will use much less than 1 page of ram.  Though useless, a good example is a counter: it's going to use at most 8-12 bytes.  Give it its own stack and it's going to take a page even though it doesn't need one.

Perhaps most coroutines are big enough that they're going to need a whole page or more all to themselves.  But it's a big cost if we assume they aren't.

I can think of ways to contrive list-like primitives that are capable of iterating over themselves and producing trait objects to each coroutine while keeping them adjacent in memory.  We don't necessarily have to pay the cost of a box.